### DISCUSSION TITLE
How to set sqlite database backend mlflow in computing cluster?

### DISCUSSION THREAD

**Discussion by Jingnan-Jia** [mlflow/mlflow#5908] [Q&A] [answered]
I use cluster (10 GPU computing nodes, 20 CPU computing nodes, 1 login nodes, 1 management node, with the shared file system) in my daily life. I found mlflow is very useful and convenient to track the experiments run on different computing nodes. But when I have more and more experiments/runs, I found that the mlflow web UI  becomes very slow. This is because the default storage method is file storage system. So the mlflow has to iterately search every directory and read the files. I think database based storage method can speed up it. So I want to set sqlite database as the backend of mlflow. Then I met some difficulties. 

If I want to use the default file storage system, I 
1. do not need to set `mlflow.set_tracking_uri()` in my python script,
2. do not need to run `mlflow server <args>` in my linux terminal,
3. use `mlflow.log_metrics()` and other log functions in my python script and then run `mlflow ui` in my linux terminal. 
Then I can successfully view the experiments records on `localhost: 5000` (I set the port local forward to view the ui in my local PC).

But if I want to use sqlite database backend, these are the steps I applied:
1. set `mlflow.set_tracking_uri()` in my python script,
2. run `mlflow server <args>` in my linux terminal,
3. use `mlflow.log_metrics()` and other log functions in my python script. 
4. Do not need to run `mlflow ui` in my linux terminal, because `mlflow server <>` has coved the function of `mlflow ui` I think.

The key issue here is how to set  `mlflow.set_tracking_uri()` in my python script, and how to set `mlflow server <args>` in my linux terminal. 

At first, I run the following line in my linux terminal on my **login node**:
`mlflow server --backend-store-uri=sqlite:///mlrunsdb.db --default-artifact-root=file:mlruns`. I hope the `Parameters`, `Metrics`, and other small files are saved in `mlrunsdb.db`, and big files like `artifacts` and `model` are saved in file directory `mlruns`.**Is my command right?**

Then, I need to set mlflow.set_tracking_uri. From the [docs](https://www.mlflow.org/docs/latest/python_api/mlflow.html#:~:text=mlflow.set_tracking_uri(uri,%5Bsource%5D), there are 4 options:
1. An empty string, or a local file path, prefixed with file:/. Data is stored locally at the provided file (or ./mlruns if empty).
2. An HTTP URI like https://my-tracking-server:5000.
3. A Databricks workspace, provided as the string “databricks” or, to use a Databricks CLI [profile](https://github.com/databricks/databricks-cli#installation), “databricks://<profileName>”.
4. A pathlib.Path instance

Now the question is how to set it in a sqlite backend mlflow in distriuted cluster? I tried several settings:
At first, do not set `mlflow.set_tracking_uri()` at all just as what I have done for the file storage system before. The result is that the code can be run successfully but the web ui did not record anything. In contrast, when I terminated the command `mlflow server --backend-store-uri=sqlite:///mlrunsdb.db --default-artifact-root=file:mlruns` using `ctrl+c`, then run `mlflow ui`, the recorded runs/experiments could be viewed now! So this means that I have to set `mlflow.set_tracking_uri()` if I want the mlflow to record parameters to database instead of default file system.

So then I tried to set uri as `mlflow.set_tracking_uri("http://localhost:5000")`.
Then I noticed that if I ran the code on different computing nodes, these experiments all failed because of the error: `ConnectionRefusedError: [Errno 111] Connection refused`. But if I ran the code on the login node, it was okay. I guess this is because the `localhost` in my `set_tracking_uri("http://localhost:5000")` means the localhost of the login node instead of any other computing nodes. **Is it right?**

So I tried to set `mlflow.set_tracking_uri("http://10.161.27.235:5000")` (Here 10.161.27.235 is the IP address of the login node). But the code can not be run successfully at all no matter on computing nodes or login node. The error is: `requests.exceptions.ConnectionError: HTTPConnectionPool(host='10.161.27.235', port=5000): Max retries exceeded with url: /api/2.0/mlflow/experiments/list?view_type=ALL (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x2aabd74e12e0>: Failed to establish a new connection: [Errno 111] Connection refused'))`

So my conclusion is that maybe I should not set HTTP URI for such a computing cluster. **Is my conclusion right?**

So I tried `mlflow.set_tracking_uri('sqlite:///mlrunsdb2.db')` (before that I ran `mlflow server --backend-store-uri=sqlite:///mlrunsdb2.db --default-artifact-root=file:mlruns` on my linux terminal in login node). 
The result was very weird. 
1. The code can be run successfully on computing node for several minute, but when I refresh the web ui or just no reason, the experiment would broke down immediatly. The error information is: `mlflow.exceptions.MlflowException: (sqlite3.OperationalError) disk I/O error (Background on this error at: https://sqlalche.me/e/14/e3q8) `
2. The code can be run successfully on login node, even refreshing the web ui can not lead it to break down. But when I launched another experiment in a computing node, the two experiments (one in login node, anoter one in computing node) will break down at the same time because of unclear reason (error is: `mlflow.exceptions.MlflowException: (sqlite3.OperationalError) disk I/O error
(Background on this error at: https://sqlalche.me/e/14/e3q8)`).

To summarize, I tried `set_tracking_uri` using HTTP URL or `sqlite:///mlrunsdb2.db`, but both failed. Errors are different. 

**Can you tell me what is the correct way to achieve sqlite database backend mlflow in computing cluster?**

**Comment by Jingnan-Jia** [comment, marked as answer]
I got the answer: 

In linux terminal, set `mlflow server --backend-store-uri=sqlite:///mlrunsdb15.db --default-artifact-root=file:mlruns --host 0.0.0.0 --port 5000`
In python script, set: `mlflow.set_tracking_uri('http://login-ip-address:5000')`

**Reply by Jingnan-Jia** [reply]
`--host 0.0.0.0` here is very important

**Reply by MoAbbasid** [reply]
@Jingnan-Jia, Hello I have a question that is a bit different from yours but Im struggling to find an answer for, Im running an ultralytics YOLOv8 model on the CLI with autologging on (Must be on CLI not python because the team is already used to using the CLI), but I did not manage to log the result to sqlite database correctly, when I run mlflow ui backend-store-uri runs/mlflow (which is a folder directory where the training logs are stored) the ui shows the results correctly, but the are saved in files inside the system at runs/mlflow path
I could not get mlflow to log to runs/mlflow.db, when I set backend-store-uri sqlite:////runs/mlflow.db a database is create at this path but is not populated with the logs which would be at the same directory.
any pointers on what to do? and please be as verbose as you can as Im trying to understand how mlflow process works.
thanks

**Reply by redoules** [reply]
> @Jingnan-Jia, Hello I have a question that is a bit different from yours but Im struggling to find an answer for, Im running an ultralytics YOLOv8 model on the CLI with autologging on (Must be on CLI not python because the team is already used to using the CLI), but I did not manage to log the result to sqlite database correctly, when I run mlflow ui backend-store-uri runs/mlflow (which is a folder directory where the training logs are stored) the ui shows the results correctly, but the are saved in files inside the system at runs/mlflow path I could not get mlflow to log to runs/mlflow.db, when I set backend-store-uri sqlite:////runs/mlflow.db a database is create at this path but is not populated with the logs which would be at the same directory. any pointers on what to do? and please be as verbose as you can as Im trying to understand how mlflow process works. thanks

Same problem here, on mlflow 2.14.2

@MoAbbasid what version are you using?

**Reply by MoAbbasid** [reply]
Hi, I solved that problem by manually setting MLFLOW_TRACKING_URI before every YOLO run, apparently it resets for some reason...
So set it before every run and check it before every run, and then run your YOLO run and it should log correctly to the sqlite db
